Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -584,6 +656,68 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
     return nil
   }
 
+  /// Visit for both the body and preamble macros.
+  func visitBodyAndPreambleMacros<Node: DeclSyntaxProtocol & WithOptionalCodeBlockSyntax>(
+    _ node: Node
+  ) -> Node {
+    // Expand preamble macros into a set of code items.
+    let preamble = expandMacros(attachedTo: DeclSyntax(node), ofType: PreambleMacro.Type.self) { attributeNode, definition in
+      expandPreambleMacro(
+        definition: definition,
+        attributeNode: attributeNode,
+        attachedTo: node,
+        in: context,
+        indentationWidth: indentationWidth
+      )
+    }
+
+    // Expand body macro.
+    let expandedBodies = expandMacros(attachedTo: DeclSyntax(node), ofType: BodyMacro.Type.self) { attributeNode, definition in
+      expandBodyMacro(
+        definition: definition,
+        attributeNode: attributeNode,
+        attachedTo: node,
+        in: context,
+        indentationWidth: indentationWidth
+      ).map { [$0] }
+    }
+
+    // Dig out the body.
+    let body: CodeBlockSyntax
+    switch expandedBodies.count {
+    case 0 where preamble.isEmpty:
+      // Nothing changes
+      return node
+
+    case 0:
+      guard let existingBody = node.body else {
+        // Any leftover preamble statements have nowhere to go, complain and
+        // exit.
+        context.addDiagnostics(from: MacroExpansionError.preambleWithoutBody, node: node)
+
+        return node
+      }
+
+      body = existingBody
+
+    case 1:
+      body = expandedBodies[0]
+
+    default:
+      context.addDiagnostics(from: MacroExpansionError.tooManyBodyMacros, node: node)
+      body = expandedBodies[0]
+    }
+
+    // If there's no preamble, swap in the new body.
+    if preamble.isEmpty {
+      return node.with(\.body, body)
+    }
+
+    var statements = body.statements
+    statements.insert(contentsOf: preamble, at: statements.startIndex)
+    return node.with(\.body, body.with(\.statements, statements))
```

Comment:
Can’t you simplify this to

```suggestion
    var node = node
    node.body.statements = preamble + body.statements
    return node
```

Reply:
`body` is optional, which makes this not directly work. But I can simplify it with `preamble + body.statements`

Reply:
Ah, I see. At least we can use `+` though 😉